isbn/alma-sru.php

```diff
@@ -0,0 +1,281 @@
+    echo "Weder isbn noch ppn Parameter für eine Suche angegeben.\n";
+    exit;
+}
+
+$suchString = '';
+
+if (isset($_GET['ppn'])) {
+    $n = trim($_GET['ppn']);
+    $searchObject = "ppn";
+}
+if (isset($_GET['isbn'])) {
+    $n = trim($_GET['isbn']);
+    $searchObject = "isbn";
+}
+$nArray = preg_split("/\s*(or|,|;)\s*/i", $n, -1, PREG_SPLIT_NO_EMPTY);
+$suchString = 'alma.all=' . implode('+OR+alma.all=', $nArray);
+$filteredSuchString = 'alma.mms_tagSuppressed=false' . '+AND+(' . $suchString . ')';
+
+# work around ExLibris server configuration issue
+# and increase timeout (i.e. waiting time)
+$contextOptions = [
+    'ssl' => [
+        'verify_peer' => true,
+        'ciphers' => 'DEFAULT@SECLEVEL=1',
+    ],
+    'http' => [
+        'timeout' => 10,
+    ],
+];
+$context = stream_context_create($contextOptions);
+$result = file_get_contents($urlBase . $filteredSuchString, false, $context);
+
+if ($result === false) {
+    header('HTTP/1.1 400 Bad Request');
+    echo "Verbindung zu SRU-Schnittstelle fehlgeschlagen\n";
+    var_dump($urlBase . $filteredSuchString);
+    exit;
+}
+
+// Delete namespaces such that we don't need to specify them
+// in every xpath query.
+$result = str_replace(' xmlns:xs="http://www.w3.org/2001/XMLSchema"', '', $result);
+$result = str_replace(' xmlns="http://www.loc.gov/MARC21/slim"', '', $result);
+
+$doc = new DOMDocument();
+$doc->preserveWhiteSpace = false;
+@$doc->loadHTML($result);
+$xpath = new DOMXPath($doc);
+
+$records = $xpath->query("//records/record/recorddata/record"); //beachte: kein CamelCase sondern alles klein schreiben
+
+$outputString = "<?xml version=\"1.0\"?>\n";
+$outputString .= "<collection>\n";
+$outputArray = [];
+
+foreach ($records as $record) {
+    // Filter out any other results which contain the ISBN but not in the 020 or 776 field
+    // or the PPN in the 001 or 035 field(s).
+    $pattern = [
+        "isbn" => './/datafield[@tag="020" or @tag="776"]/subfield',
+        "ppn" => './/controlfield[@tag="001"]|.//datafield[@tag="035"]/subfield'
+    ];
+    $foundMatch = false;
+    $nodes = $xpath->query($pattern[$searchObject], $record);
+    foreach ($nodes as $foundNode) {
+        $foundValue = $foundNode->nodeValue;
+        foreach ($nArray as $queryValue) {
+            $testString = $queryValue;
+            if ($searchObject == "isbn") {
+                $testString = preg_replace('/[^0-9xX]/', '', $testString);
+                $foundValue = preg_replace('/[^0-9xX]/', '', $foundValue);
+                if (strlen($testString) == 13) {
+                    // Delete the 978-prefix and the check value at the end for ISBN13
+                    $testString = substr($testString, 3, -1);
+                } elseif (strlen($testString) == 10) {
+                    // Delete check value at the end for ISBN10
+                    $testString = substr($testString, 0, -1);
+                }
+            }
+            if (strpos($foundValue, $testString) !== false) {
```

Comment:
Vielleicht sollten wir für ID hier eine exakte Übereinstimmung fordern (sonst Gefahr eine ID nach Schema X wird auch in Schema Y als Teil genutzt aber gehört zu einem ganz anderen Title). Aber dann muss man die jeweiligen Präfixe in den 035er Felder auch eingeben.

Reply:
Beispiel für weitere IDs in 035er Felder:
```xml
<datafield tag="035" ind1="" ind2=""><subfield code="a">(MiAaPQ)EBC31745127</subfield></datafield>
<datafield tag="035" ind1="" ind2=""><subfield code="a">(Au-PeEL)EBL31745127</subfield></datafield>
<datafield tag="035" ind1="" ind2=""><subfield code="a">(CaSebORM)9783960092476</subfield></datafield>
<datafield tag="035" ind1="" ind2=""><subfield code="a">(EXLCZ)995860000000545454</subfield></datafield>
```
Suche nach dem ganze Wert, z.B. "(MiAaPQ)EBC31745127" funktioniert nicht, da die Alma-SRU-Schnittstelle dann gar nichts zurückliefert. Daher ist das Beste doch nur den Teil nach der Klammer zu suchen, aber die Prüfung kann man stärker machen (bis auf den Klammerzusatz müssen die Werte gleich sein).